uniffi_bindgen/src/bindings/gecko/templates/SharedHeaderTemplate.h

```diff
@@ -0,0 +1,795 @@
+};
+
+{% for func in ci.iter_ffi_function_definitions() -%}
+{%- match func.return_type() -%}
+{%- when Some with (type_) %}
+{{ type_|type_ffi }}
+{% when None %}
+void
+{%- endmatch %}
+{{ func.name() }}(
+    {%- for arg in func.arguments() %}
+    {{ arg.type_()|type_ffi }} {{ arg.name() -}}{%- if loop.last -%}{%- else -%},{%- endif -%}
+    {%- endfor %}
+    {%- if func.has_out_err() -%}{%- if func.arguments().len() > 0 %},{% endif %}
+    RustError* out_err
+    {%- endif %}
+);
+
+{% endfor -%}
+
+}  // extern "C"
+
+namespace mozilla {
+namespace dom {
+// namespace {{ ci.namespace() }} {
+
+// TODO: Rename this to something less conflict-y, and
+// make sure `lift_cpp` and `lower_cpp` know about it.
+namespace detail {
+
+/// A helper class to read values out of a Rust byte buffer.
+class MOZ_STACK_CLASS Reader final {
+ public:
+  explicit Reader(const RustBuffer& aBuffer) : mBuffer(aBuffer), mOffset(0) {}
+
+  /// Indicates if the offset has reached the end of the buffer.
+  bool HasRemaining() {
+    return static_cast<int64_t>(mOffset.value()) < mBuffer.mLen;
+  }
+
+  /// Helpers to read fixed-width primitive types at the current offset.
+  /// Fixed-width integers are read in big endian order.
+
+  uint8_t ReadUInt8() {
+    return ReadAt<uint8_t>(
+        [this](size_t aOffset) { return mBuffer.mData[aOffset]; });
+  }
+
+  int8_t ReadInt8() { return BitwiseCast<int8_t>(ReadUInt8()); }
+
+  uint16_t ReadUInt16() {
+    return ReadAt<uint16_t>([this](size_t aOffset) {
+      uint16_t value;
+      memcpy(&value, &mBuffer.mData[aOffset], sizeof(uint16_t));
+      return PR_ntohs(value);
+    });
+  }
+
+  int16_t ReadInt16() { return BitwiseCast<int16_t>(ReadUInt16()); }
+
+  uint32_t ReadUInt32() {
+    return ReadAt<uint32_t>([this](size_t aOffset) {
+      uint32_t value;
+      memcpy(&value, &mBuffer.mData[aOffset], sizeof(uint32_t));
+      return PR_ntohl(value);
+    });
+  }
+
+  int32_t ReadInt32() { return BitwiseCast<int32_t>(ReadUInt32()); }
+
+  uint64_t ReadUInt64() {
+    return ReadAt<uint64_t>([this](size_t aOffset) {
+      uint64_t value;
+      memcpy(&value, &mBuffer.mData[aOffset], sizeof(uint64_t));
+      return PR_ntohll(value);
+    });
+  }
+
+  int64_t ReadInt64() { return BitwiseCast<int64_t>(ReadUInt64()); }
+
+  float ReadFloat() { return BitwiseCast<float>(ReadUInt32()); }
+
+  double ReadDouble() { return BitwiseCast<double>(ReadUInt64()); }
+
+  /// Reads a length-prefixed UTF-8 encoded string at the current offset. The
+  /// closure takes a `Span` pointing to the raw bytes, which it can use to
+  /// copy the bytes into an `nsCString` or `nsString`.
+  ///
+  /// Safety: The closure must copy the span's contents into a new owned string.
+  /// It must not hold on to the span, as its contents will be invalidated when
+  /// the backing Rust byte buffer is freed. It must not call any other methods
+  /// on the reader.
+  template <typename T>
+  void ReadRawString(
+      T& aString,
```

Comment:
Using an out parameter here because `AString` types (`nsACString` and `nsAString`) can’t be returned by value.

Reply:
This seems like a good comment to put in a comment :)